src/umpire/util/io.cpp

```diff
@@ -0,0 +1,181 @@
+#include <fstream>
+#include <ostream>
+#include <stdlib.h>   // for getenv()
+
+#if defined(UMPIRE_ENABLE_FILESYSTEM)
+#include <filesystem>
+#else
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <dirent.h>
+#endif
+
+#if !defined(_MSC_VER)
+#include <unistd.h>   // getpid()
+#else
+#include <process.h>
+#define getpid _getpid
+#endif
+
+namespace umpire {
+
+static util::OutputBuffer s_log_buffer;
+static util::OutputBuffer s_replay_buffer;
+static util::OutputBuffer s_error_buffer;
+
+std::ostream log{&s_log_buffer};
+std::ostream replay{&s_replay_buffer};
+std::ostream error{&s_error_buffer};
+
+namespace util {
+
+static std::string make_unique_filename(
+  const std::string& base_dir,
+  const std::string& name,
+  const int pid,
+  const std::string& extension);
+
+static inline bool file_exists(const std::string& file);
+
+static inline bool directory_exists(const std::string& file);
+
+void initialize_io(const bool enable_log, const bool enable_replay)
+{
+  s_log_buffer.setConsoleStream(&std::cout);
+  s_replay_buffer.setConsoleStream(nullptr);
+  s_error_buffer.setConsoleStream(&std::cerr);
+
+  std::string root_io_dir{"./"};
+  const char* output_dir{std::getenv("UMPIRE_OUTPUT_DIR")};
+  if (output_dir) root_io_dir = output_dir;
+
+  std::string file_basename{"umpire"};
+  const char* base_name{std::getenv("UMPIRE_OUTPUT_BASENAME")};
+  if (base_name) file_basename = base_name;
+
+  const int pid{getpid()};
+
+  const std::string log_filename{
+    make_unique_filename(root_io_dir, file_basename, pid, "log")};
+
+  const std::string replay_filename{
+    make_unique_filename(root_io_dir, file_basename, pid, "replay")};
+
+  const std::string error_filename{
+    make_unique_filename(root_io_dir, file_basename, pid, "error")};
+
+  if (!directory_exists(root_io_dir)) {
+    if (MPI::isInitialized()) {
+      if (MPI::getRank() == 0) {
+#if defined(UMPIRE_ENABLE_FILESYSTEM)
+        std::filesystem::path root_io_dir_path{root_io_dir};
+
+        if (!std::filesystem::exists(root_io_dir_path) &&
+            (enable_log || enable_replay)) {
+          std::filesystem::create_directories(root_io_dir_path);
+        }
+#else
+        struct stat info;
+        if ( stat( root_io_dir.c_str(), &info ) )
+        {
+          if (enable_log || enable_replay) {
+            if ( mkdir(root_io_dir.c_str(),
+                       S_IRWXU | S_IRGRP | S_IXGRP | S_IROTH | S_IXOTH) ) {
+              UMPIRE_ERROR("mkdir(" << root_io_dir << ") failed");
+            }
+          }
+        }
+        else if ( !(S_ISDIR(info.st_mode)) )
+        {
+          UMPIRE_ERROR(root_io_dir << "exists and is not a directory");
+        }
+#endif
+      }
+      MPI::sync();
+    } else {
+      UMPIRE_ERROR("Cannot create output directory before MPI has been initialized. "
+                   "Please unset UMPIRE_IO_DIR in your environment");
+    }
+  }
+
+  if (enable_log) {
+    static std::ofstream s_log_ofstream{log_filename};
```

Comment:
I moved `s_log_ofstream` and `s_replay_ofstream` inside this scope. Let me know if you see a reason to move them for greater access.